File: C++/Sort_Quick.cpp
/*
 * C++ PROGRAMMING EXAMPLES EXAMPLES
 * 
 * This file demonstrates C++ programming concepts and techniques:
 * - Core concepts and syntax
 * - Practical examples and usage patterns
 * - Best practices and optimization techniques
 * - Real-world applications and scenarios
 * 
 * Learning Objectives:
 * - Understand fundamental C++ PROGRAMMING concepts
 * - Learn proper syntax and implementation
 * - Master different techniques and approaches
 * - Apply knowledge in practical scenarios
 * - Develop problem-solving skills
 * 
 * Prerequisites:
 * - Basic C++ knowledge
 * - Understanding of data types and variables
 * - Familiarity with control structures
 * - Basic understanding of functions and classes
 * 
 * Key Topics Covered:
 * - Syntax and language features
 * - Implementation techniques
 * - Performance considerations
 * - Error handling and best practices
 * - Real-world applications
 */

/* Quick Sort 
Time Complexity of Quick Sort
Case	Time Complexity
Best Case	O(n log n) → Balanced partitions
Average Case	O(n log n) → Random inputs
Worst Case	O(n²) → Already sorted or reversed (bad pivot)

Space Complexity
O(log n) → Recursion stack

https://visualgo.net/en/sorting
*/

#include <iostream>
using namespace std;

// Function to partition the array and return the pivot index
int partition(int arr[], int low, int high) {
    int pivot = arr[high]; // Choosing the last element as the pivot
    int i = low - 1; // Index of smaller element

    for (int j = low; j < high; j++) {
        // If current element is less than or equal to pivot
        if (arr[j] <= pivot) {
            i++;
            // Swap arr[i] and arr[j]
            swap(arr[i], arr[j]);
        }
    }

    // Swap pivot to correct position
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Recursive Quick Sort function
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        // pivot is partition index, arr[pi] is at right place
        int pi = partition(arr, low, high);

        // Recursively sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Function to print the array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++)
        cout << arr[i] << " ";
    cout << endl;
}

// Main to test
int main() {
    int arr[] = {10, 7, 8, 9, 1, 5};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Original array: ";
    printArray(arr, n);

    quickSort(arr, 0, n - 1);

    cout << "Sorted array: ";
    printArray(arr, n);

    return 0;
}

File: python_old/sort_quick.py
'''
Quick Sort 
Time Complexity of Quick Sort
Case    Time Complexity
Best Case   O(n log n) → Balanced partitions
Average Case    O(n log n) → Random inputs
Worst Case  O(n²) → Already sorted or reversed (bad pivot)

Space Complexity
O(log n) → Recursion stack

https://visualgo.net/en/sorting
'''

def quick_sort(arr):
    if len(arr) <= 1:
        return arr  # Base case: already sorted

    pivot = arr[len(arr) // 2]  # Choose the middle element as pivot
    left = [x for x in arr if x < pivot]       # Elements less than pivot
    # print('Left: ', left)
    middle = [x for x in arr if x == pivot]    # Elements equal to pivot
    # print('Middle: ', middle)
    right = [x for x in arr if x > pivot]      # Elements greater than pivot
    # print('Right: ', right)

    # Recursively sort left and right, then combine
    return quick_sort(left) + middle + quick_sort(right)

# Example usage:
arr = [47, 2, 10, 7, 22, 8, 9, 1, 13, 5, 7, 4, 32, 6]
print("Original array:", arr)
sorted_arr = quick_sort(arr)
print("Sorted array:  ", sorted_arr)